### PR TITLE
resolve #6670 add back support for search --reverse-dependency

### LINKED ISSUE #6670: conda search --reverse-dependency not working on conda 4.4
The `--reverse-dependency` argument does not seem to effect the results of `conda search` in conda 4.4:

```
~$ conda search --reverse-dependency networkx
Loading channels: done
Name                       Version                   Build  Channel        
networkx                   1.7                      py26_0  defaults       
networkx                   1.7                      py27_0  defaults       
networkx                   1.7                      py33_0  defaults       
networkx                   1.8                      py26_0  defaults       
networkx                   1.8                      py27_0  defaults
...

~$ conda search --reverse-dependency networkx
Fetching package metadata ...........
anaconda                     1.6.0                np16py26_0  defaults        
                             1.6.0                np16py27_0  defaults        
                             1.6.0                np17py26_0  defaults        
                             1.6.0                np17py27_0  defaults        
                             1.6.0                np17py33_0  defaults        
                             1.6.1                np16py26_0  defaults        
                             1.6.1                np16py27_0  defaults        
                             1.6.1                np17py26_0  defaults        
                             1.6.1                np17py27_0  defaults        
                             1.6.1                np17py33_0  defaults
...
```

### PR DESCRIPTION
**Pull request by kalefranz**
resolve #6670

### CODE REVIEW

**Comment by kalefranz** [comment]
@jjhelmus Since you did this initial work here, do you want to review this PR?

**Comment by jjhelmus** [comment]
Thanks for moving this forward @kalefranz.  I'll review this later today.

**Comment by jjhelmus** [comment]
This works but the requirements requiring a single package make using this unappealing.  The old regex based behavior did not have this limitation.  In addition the matching does not appear to take into account the details of the single package. 

```
$ python -m "conda" search --reverse-dependency automat
Loading channels: done
# Name                  Version           Build  Channel             
automat                   0.5.0          py27_0  pkgs/free           
automat                   0.5.0          py35_0  pkgs/free           
automat                   0.5.0          py36_0  pkgs/free           
automat                   0.6.0  py27hfed452d_0  pkgs/main           
automat                   0.6.0  py35h367829d_0  pkgs/main           
automat                   0.6.0  py36ha140c00_0  pkgs/main           

MultiplePackagesError: Multiple packages found for spec: automat
Please refine your query to result in a single package to reverse from.
```

With conda 4.3.x this would find twisted without the need to further refine the package spec.

With a single package specification:

```
$ python -m "conda" search --reverse-dependency automat=0.5.0=py27_0
Loading channels: done
# Name                  Version           Build  Channel             
twisted                  17.1.0          py27_0  pkgs/free           
twisted                  17.1.0          py35_0  pkgs/free           
twisted                  17.1.0          py36_0  pkgs/free           
twisted                  17.5.0          py27_0  pkgs/free           
twisted                  17.5.0          py35_0  pkgs/free           
twisted                  17.5.0          py36_0  pkgs/free           
twisted                  17.9.0  py27h7008fb7_0  pkgs/main           
twisted                  17.9.0  py35h812867c_0  pkgs/main           
twisted                  17.9.0  py36hb8d8d5a_0  pkgs/main
```

This seems odd since the Python 2.7 package of automat cannot fulfill the dependencies of the python 3.5 and 3.6 twisted packages.

File: conda/cli/conda_argparse.py [hunk 1]
```diff
@@ -918,9 +918,8 @@ def configure_parser_search(sub_parsers):
     p.add_argument(
         "--reverse-dependency",
         action="store_true",
-        help="Perform a reverse dependency search. When using this flag, the --full-name "
-             "flag is recommended. Use 'conda info package' to see the dependencies of a "
-             "package.",
+        help="Perform a reverse dependency search. Use 'conda search package --info' to see the "
+             "dependencies of a package.",
```
Comment (goanpeca) [inline conda/cli/conda_argparse.py]: Should we be explicit about the 1-level-only search?

**Comment by kalefranz** [comment]
I don't think we're at the behavior we want yet.  Maybe the first thing to do in this case is to write the tests, with the output that we expect.  Make the tests touch some of the more complicated corner cases.  And then we can figure out how to implement it.

**Comment by kalefranz** [comment]
@jjhelmus A new implementation for you to review, as discussed on flowdock.

**Comment by jjhelmus** [comment]
The automat example from above does not seem to give the expected results

```
$ python -m conda search --reverse-dependency automat
Searching: done

PackagesNotFoundError: The following packages are not available from current channels:

  - automat

Current channels:

  - https://repo.anaconda.com/pkgs/main/linux-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/free/linux-64
  - https://repo.anaconda.com/pkgs/free/noarch
  - https://repo.anaconda.com/pkgs/r/linux-64
  - https://repo.anaconda.com/pkgs/r/noarch
  - https://repo.anaconda.com/pkgs/pro/linux-64
  - https://repo.anaconda.com/pkgs/pro/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.
```
I would expect this to return various twisted packages.

These seems to be a result of the automat dependency in twisted being constrained, `automat >=0.3.0`.

**Comment by jjhelmus** [comment]
Adding a version to the package gives the expected results but I do not think this should be required:

```
$ python -m conda search --reverse-dependency automat==0.3.0
Searching: done
# Name                  Version           Build  Channel             
twisted                  17.1.0          py27_0  pkgs/free           
twisted                  17.1.0          py35_0  pkgs/free           
twisted                  17.1.0          py36_0  pkgs/free           
twisted                  17.5.0          py27_0  pkgs/free           
twisted                  17.5.0          py35_0  pkgs/free           
twisted                  17.5.0          py36_0  pkgs/free           
twisted                  17.9.0  py27h7008fb7_0  pkgs/main           
twisted                  17.9.0  py35h812867c_0  pkgs/main           
twisted                  17.9.0  py36hb8d8d5a_0  pkgs/main   
```

**Comment by kalefranz** [comment]
Right. This implementation is now wrong.  We're back to trying to combine fuzzy matches I think, and I just don't know how to do that without a substantial amount of logic.  The problem is fundamentally that `.match()` only goes one way.  We can't match a spec to a spec.  Only a spec to an exact description of a package.

**Comment by jjhelmus** [comment]
Perhaps implementing reverse-dependency search as a regex is the best option for the time being.

**Comment by kalefranz** [comment]
Try that.

**Comment by mcg1969** [comment]
Can we reprioritize this? I'm disappointed this was removed—it's one of those things you don't need often but when you do...

**Comment by wolfv** [comment]
we're working on this kind of functionality in mamba. We currently have a repoquery feature that allows to query for forward and reverse dependencies and displays them in a tree or table.

However, we've artificially restricted the feature to installed packages. I've tried the automat example with reverse dependency and it works in mamba, but not when specifying relational operators (e.g.. `mamba repoquery whoneeds automat`) shows a bunch of twisted versions etc.
I need to figure out how to make it work with `automat >=0.5`. If there is nothing pre-made in libsolv for this, I believe it would be enough to find all reverse dependencies without selector, and then check if the matchspec intersection yields a result?

**Comment by travishathaway** [comment]
This pull request has become stale and is therefore being closed.
